### PR TITLE
IGNITE-19410 Fixed node crash due to SecurityContext not being found during discovery message processing.

### PR DESCRIPTION
**Pull request by petrov-mg**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/NodeSecurityContextPropagationTest.java [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+            getTestTimeout());
+    }
+
+    /** */
+    private IgniteEx startClientNode(int idx) throws Exception {
+        return startGrid(getConfiguration(getTestIgniteInstanceName(idx), true));
+    }
+
+    /** */
+    private boolean isNodeAddedMessageReceived(IgniteEx ignite, int nodeIdx) {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/NodeSecurityContextPropagationTest.java]: I refactored the `getReceivedMessages` method to use `Predicate` as its parameter. This made the test logic a bit clearer. But a separate method for checking whether an arbitrary message is a DiscoveryNodeAddedMessage message sill remains.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/NodeSecurityContextPropagationTest.java [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        discoveryMessageQueue(srv).unblock();
+
+        waitForCondition(
+            () -> grid(0).cluster().nodes().size() == 4
+                && TEST_MESSAGE_ACCEPTED_NODES.contains(grid(2).cluster().localNode().id()),
+            getTestTimeout());
+    }
+
+    /** */
+    private IgniteEx startClientNode(int idx) throws Exception {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/NodeSecurityContextPropagationTest.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/NodeSecurityContextPropagationTest.java [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return getConfiguration(igniteInstanceName, false);
+    }
+
+    /** */
+    private IgniteConfiguration getConfiguration(String igniteInstanceName, boolean isClient) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName)
+            .setFailureHandler(new StopNodeOrHaltFailureHandler())
+            .setDataStorageConfiguration(new DataStorageConfiguration())
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/security/NodeSecurityContextPropagationTest.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 4]
```diff
@@ -2678,6 +2678,13 @@ public ClusterNode historicalNode(UUID nodeId) {
                 return node;
         }
 
+        for (Collection<ClusterNode> top : topHist.values()) {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java]: Fixed.

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**